### ISSUE TITLE
Quality plugin not working

### ISSUE THREAD

**Issue by throne1986** [mediaelement/mediaelement-plugins#172]
Hi every one am new to media element , I would like to use quality plugin but  unfortunately its not showing here is my code

    <!DOCTYPE html>
    <html lang="en">
        <head>
            <meta charset="UTF-8" />
            <meta name="viewport" content="width=device-width, initial-scale=1.0" />
            <title>Demo HLS</title>
            <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/5.0.0/normalize.min.css" />
            <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/mediaelement/4.2.6/mediaelementplayer.css" />
            <link rel="stylesheet" href="https://github.com/mediaelement/mediaelement-plugins/blob/master/dist/quality/quality.css" />
            <link rel="icon" href="https://www.interactive-video.videomill.pl/images/fav_videomill_1.png" />
        </head>
        <body>
            <div class="media-wrapper">
                <video id="player1" width="750" height="421" controls preload="none">
                    <source type="application/x-mpegURL" src="//r.dcs.redcdn.pl/hls/o2/Allegro/vod/1af1b7ee-f87b-4e55-b50f-ac9d7f3e2647/playlist.smil/playlist.m3u8" data-quality="HD" />
                    <source type="video/mp4" src="https://commondatastorage.googleapis.com/gtv-videos-bucket/CastVideos/mp4/BigBuckBunny.mp4" data-quality="SD" />
                    <source type="video/mp4" src="http://clips.vorwaerts-gmbh.de/big_buck_bunny.mp4" data-quality="LD" />
                </video>
            </div>
    
            <script src="https://cdnjs.cloudflare.com/ajax/libs/mediaelement/4.2.6/mediaelement-and-player.min.js"></script>
            <script src="https://github.com/mediaelement/mediaelement-plugins/blob/master/dist/quality/quality.js"></script>
    
            <script>
                var mediaElements = document.querySelectorAll("video, audio");
    
                for (var i = 0, total = mediaElements.length; i < total; i++) {
                    new MediaElementPlayer(mediaElements[i], {
                        features: ["playpause", "current", "progress", "duration", "volume", "quality", "fullscreen"],
                    });
                }
            </script>
        </body>
    </html>

> What is wrong with my code? u can copy the above code and run it , there is no quality button at all

**Comment by jaapmarcus** [comment]
Prepared a Codepen for you see maybe it will help you :)
https://codepen.io/jaapmarcus/pen/GRpxZWj
Best regards

**Comment by throne1986** [comment]
@jaapmarcus  thanks a lot, is it possible to play video based on internet speed? let assume my internet is very slow when I open the media player automatically it should play low-quality video eg sd ?

**Comment by jaapmarcus** [comment]
The only way to do that is with downloading a file and testing it. Not the best option imo..

HLS and Dash protocol / steaming has build in support for it.  

If you check https://r.dcs.redcdn.pl/hls/o2/Allegro/vod/1af1b7ee-f87b-4e55-b50f-ac9d7f3e2647/playlist.smil/playlist.m3u8 with out any player you will see it scaling depending on network performance. You will need to supply multiple stream qualities in 1 file ima...

**Comment by throne1986** [comment]
Yes Ur right everything works perfectly now, my last question is about autoplay, I added autoplay true on media element player options but it's not working, do I always need to do like this below?

    <video id="videoplayer" muted muted webkit-playsinline playsinline autoplay>
       </video>
or is there any other method?

**Comment by jaapmarcus** [comment]
```
<video id="videoplayer" muted playsinline autoplay>
   </video>

```
Should be fine 

But yes it needs to be always added for iOS other mobile devices. (And some other browsers)

**Comment by throne1986** [comment]
okay thanks, bro u can close the issue, by the way, this amazing player, for me the best open source available in the market, God Bless all yah for the amazing job

**Comment by mircobabini** [comment]
Hey, adding some more questions here:

1) About HLS, what do you think about preparing a demo with mediaelement + https://github.com/aminyazdanpanah/PHP-FFmpeg-video-streaming#hls ? I think that a lot of people could be in my same situation (I just have 1 high quality mp4 and want to generate medium and low quality resources and also add HLS in order to have auto-quality selection, but don't know how to create the m3u8 files you linked above @jaapmarcus).

2) About redundancy, is it possible to have the same resources provided by multiple servers, in order to have an automatic fallback in case one of the resources goes in 404/40* error for any reason?

Thank you.

**Comment by jaapmarcus** [comment]
> 1. About HLS, what do you think about preparing a demo with mediaelement + https://github.com/aminyazdanpanah/PHP-FFmpeg-video-streaming#hls ? I think that a lot of people could be in my same situation (I just have 1 high quality mp4 and want to generate medium and low quality resources and also add HLS in order to have auto-quality selection, but don't know how to create the m3u8 files you linked above @jaapmarcus).

It is possible with FFmpeg see https://docs.peer5.com/guides/production-ready-hls-vod/

In the past I made a Wordpress  Plugin based on the transcoding (mp4 to a lower quality mp4) with PHP-ffmpeg so allot is possible. I suggest at least do the encoding one time and save it to the disk... On the fly encoding is only possible with low traffic or you need a way bigger server to enanble it.

Also paid options are 3rd parties like https://coconut.co, or AWS and stuff like that.

> 2. About redundancy, is it possible to have the same resources provided by multiple servers, in order to have an automatic fallback in case one of the resources goes in 404/40* error for any reason?

This is possible with a Propper Loadbalancer and multiple servers. How ever this is way beyond the scope of Media Elements...

**Comment by mircobabini** [comment]
Sure, of course it's beyond the scope, I'll do my own researches. Just one more question: in order to use BLOB urls instead of standard urls the best idea is to first enable HLS in order to activate segmentation instead of having to download the full mp4 video synchronously, right?

Thank you very much.

**Comment by jaapmarcus** [comment]
Go ahead and try. I don't using blobs work due to the nature of HLS and the segmented part.

**Comment by vendbits** [comment]
> Prepared a Codepen for you see maybe it will help you :)
> https://codepen.io/jaapmarcus/pen/GRpxZWj
> Best regards

Hello,

I'm having an issue with the quality plugin also as described here: https://github.com/mediaelement/mediaelement-plugins/issues/196

I looked at the Codepen mentioned above and I noticed it has the same behavior. The video source/quality is not changing as I can see in Inspect Element > Network

Only **BigBuckBunny.mp4**, and never **big_buck_bunny.mp4** is being played even after switching source/quality

Thank you.

**Comment by rotator** [comment]
Hi!

I had the same bug with both versions: mediaelement-plugins 2.5.0 + mediaelement 4.2.17 and mediaelement-plugins 2.6.0 + mediaelement 5.0.0.

The problem was in the newQuality value after radio input change - lines 224-225 in quality.js. 
Fixed it for the latest releases mediaelement-plugins 2.6.0 + mediaelement 5.0.0 and created a patch (please, remove .md extension) [mediaelement-plugins-quality-switcher.patch.md](https://github.com/mediaelement/mediaelement-plugins/files/7236201/mediaelement-plugins-quality-switcher.patch.md).

**Comment by jaapmarcus** [comment]
Please submit a pull request via https://github.com/mediaelement/mediaelement-plugins/pulls

**Comment by rotator** [comment]
Updated fix for the mediaelement-plugins 2.6.2 release. Tested with mediaelement 5.0.4 version.
Created PR https://github.com/mediaelement/mediaelement-plugins/pull/215
Please, review.
